Adding capability for charged particle motion in electric and magnetic fields

gsgall opened the pull request:
- VelocityUpdater was refactored to ParticleStepper to be more descriptive
- The updateVelocity method was changed to setupStep since it does more than just update the particle velocity

(Closes #19)

gsgall committed:
Closes #19: Adding capability for charged particle motion in electric and magnetic fields

- VelocityUpdater was refactored to ParticleStepper to be more descriptive
- The updateVelocity method was changed to setupStep since it does more than just update the particle velocity

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2145068/) on 1ca808c wanted to post the following:

View the site [here](https://fenix-dev.hpc.inl.gov/PRs/20/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2145072/) on 1ca808c wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>042c1a</th>
      <th colspan=4>#20 1ca808</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.53%</td>
      <td>91.90%</td>
      <td>+0.38%</td>
      <td>94.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>162</td>
      <td>227</td>
      <td>+65</td>
      <td>94</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15</td>
      <td>20</td>
      <td>+5</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://fenix-dev.hpc.inl.gov/PRs/20/coverage_diff)

[Full coverage report](https://fenix-dev.hpc.inl.gov/PRs/20/coverage)



This comment will be updated on new commits.

gsgall committed:
removing banana

simopier commented on doc/content/bib/boris.bib:
```diff
@@ -0,0 +1,24 @@
+@article{qin2013boris,
```
Could you rename this bib file? `boris.bib` is very specific, and we probably want something more general that would contain all the references. Maybe `references.bib`?

gsgall replied:
This file has been renamed to references.bib

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+{
+  auto params = ParticleStepperBase::validParams();
+  params.addClassDescription(
+      "Electromagnetic particle stepper method implementing the Boris Algorithm");
```
```suggestion
      "Electromagnetic particle stepper method implementing the Boris Algorithm.");
```

simopier commented on doc/content/source/userobjects/BorisStepper.md:
```diff
@@ -0,0 +1,111 @@
+  \frac{q}{m}
+  \vec{E}_n
+  \frac{\Delta t}{2}
+\end{equation}
```
The documentation should follow the same guidance provided in the paper: equations need to be part of sentences, notations need to be defined, etc. 

I reproduced the content of the paper with some additional links to other documentation page and test input files. 
By the way, do you have a test for the constant electric field? You could add a link to it like I did for the other two tests. 


```suggestion
In magnetized (or electromagnetic) PIC simulations, the de facto standard particle stepping algorithm is commonly known as the Boris algorithm [!cite](boris1970relativistic,Birdsall_Langdon_1991, qin2013boris). This algorithm is similar to a [Leap Frog method](userobjects/LeapFrogStepper.md) and has second order accuracy in time when solving the equations of motion for a charged particle, given by
\begin{equation} \label{eq:pos}
  \diff{\vec{r}}{t}
  =
  \vec{v}
\end{equation}
and 
\begin{equation} \label{eq:vel}
  \diff{\vec{v}}{t}
  =
  \frac{q}{m}
  \left[
    \vec{E}
    +
    \vec{v}
    \times
    \vec{B}
  \right]
\end{equation}
where $q$ is the particle's charge, $m$ is the particle's mass, and  $\vec{E}$ and $\vec{B}$ are the electric and magnetic fields that the particle is subject to, respectively.

In the Boris algorithm, \cref{eq:pos,eq:vel} are discretized with a central difference scheme and the acceleration due to the electric field and magnetic field are separated. First, half of the impulse due to the electric field is applied to the particle, as  
\begin{equation} \label{eq:e_half1}
  \vec{v}^{\,-}
  =
  \vec{v}_{n}
  +
  \frac{q}{m}
  \vec{E}_n
  \frac{\Delta t}{2}
\end{equation}
where $\vec{v}^{\,-}$ is an intermediate particle velocity, $\vec{v}_{n}$ is the particle velocity at step $n$, and $\vec{E}_{n}$ is the electric field at step $n$. The velocity of the particle after rotation due to the magnetic field is derived as
\begin{equation} \label{eq:boris_v_plus}
  \vec{v}^{\,+}
  =
  \vec{v}^{\,-}
  +
  \vec{v}^{\,'}
  \times
  \vec{s},
\end{equation}
with 
\begin{equation} \label{eq:boris_v_prime}
  \vec{v}^{\,'}
  =
  \vec{v}^{\,-}
  +
  \vec{v}^{\,-}
  \times
  \vec{l}
\end{equation}
where 
\begin{equation} \label{eq:boris_t}
  \vec{l} =
  \frac{q}{m}
  \vec{B}_n
  \Delta t,
\end{equation}
which accounts for the effect of $\vec{B}_n$, the magnetic field at step $n$. $\vec{s}$ is defined as
\begin{equation} \label{eq:boris_s}
  \vec{s} =
  \frac{2 \vec{l}}{
    1 + \vec{l} \cdot \vec{l}
  }.
\end{equation}
Finally, the rotation due to the presence of the magnetic field is then applied with
\begin{equation} \label{eq:mag_step}
  \frac{
    \vec{v}^{\,+}
    -
    \vec{v}^{\,-}
  }{ \Delta t}
  =
  \frac{q}{m}
  \left(
    \vec{v}^{\,+}
    +
    \vec{v}^{\,-}
  \right)
  \times
  \vec{B}_n,
\end{equation}
and the final impulse due to the electric field is then applied to the particle using 
\begin{equation} \label{eq:e_half2}
  \vec{v}_{n+1}
  =
  \vec{v}^{\,+}
  +
  \frac{q}{m}
  \vec{E}_n
  \frac{\Delta t}{2}.
\end{equation}
The implementation of the Boris algorithm was verified using several single particle motion tests: constant electric field, cyclotron motion ([/cyclotron_motion.i]), and $\vec{E} \times \vec{B}$ drift motion ([/e_cross_b.i]).
```

simopier commented on doc/content/source/userobjects/BorisStepper.md:
```diff
@@ -0,0 +1,111 @@
+
+More details about the Boris algorithm can be found in
+
+!bibtex list bib_files=boris.bib
```
```suggestion
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+
+!syntax description /UserObjects/LeapFrogStepper
+
+On the first step that a particle takes the particle velocity is updated with a 1/2 time step
```
```suggestion
On the first step that a particle takes the particle velocity is updated with a 1/2 time step as
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+
+\begin{equation}
+  \vec{v}_{1/2} = \vec{v}_0 + \frac{q}{m} \vec{F} \frac{dt}{2}
+\end{equation}
```
please explain all the notations (see comments provided in `BorisStepper.md`)
```suggestion
\begin{equation}
  \vec{v}_{1/2} = \vec{v}_0 + \frac{q}{m} \vec{F} \frac{dt}{2},
\end{equation}
where ...
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+  \vec{v}_{1/2} = \vec{v}_0 + \frac{q}{m} \vec{F} \frac{dt}{2}
+\end{equation}
+
+The particle position is then updated with a full time step
```
```suggestion
The particle position is then updated with a full time step as
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+
+The particle position is then updated with a full time step
+
+\begin{equation}
+  \vec{r}_{1} = \vec{r}_0 + \vec{v}_{1/2} dt
+\end{equation}
```
```suggestion
\begin{equation}
  \vec{r}_{1} = \vec{r}_0 + \vec{v}_{1/2} dt,
\end{equation}
with ....
```

simopier replied:
Again, please describe notations.

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+  \vec{r}_{1} = \vec{r}_0 + \vec{v}_{1/2} dt
+\end{equation}
+
+For every other time step, n, after the initial both are updated with a full time step and
```
```suggestion
For every other time step, $n$, after the initial both are updated with a full time step and
```
Are you missing a word in ` after the initial both` ?

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+
+\begin{equation}
+  \vec{r}_{n} = \vec{r}_{n-1} + \vec{v}_{n + 1/2} dt
+\end{equation}
```
Make sure to incorporate these equations in a sentence and define the notations that have not yet been defined.

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+# LeapFrogStepper
```
Do you have references for this approach? Please add it to this page. You can also add your submitted paper on this page and the BorisStepper page, we'll just update it once it's published.

gsgall replied:
There is some more detail about this in the Birdsall Langdon book and it has been added as a reference

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,35 @@
+# LeapFrogStepper
+
+!syntax description /UserObjects/LeapFrogStepper
+
```
Could you please add more context here? Why is this method important? What is it used for?

gsgall replied:
More context has been added

simopier commented on doc/content/source/userobjects/BorisStepper.md:
```diff
@@ -0,0 +1,111 @@
+
+!syntax inputs /UserObjects/BorisStepper
+
+!syntax children /UserObjects/BorisStepper
```
```suggestion
!syntax children /UserObjects/BorisStepper

!bibtex bibliography
```

simopier commented on doc/content/source/userobjects/ParticleStepperBase.md:
```diff
@@ -0,0 +1,9 @@
+# ParticleStepperBase
+
+`ParticleStepperBase` is the abstract class form which all steppers should be derrived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the veloctiy stored in ray data, the current time step and dimension of the mesh. Addtionally, several other useful methods are provided. In the provided method `setMaxDistanceAndDirection` the particl ...
```
```suggestion
`ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the velocity stored in ray data, the current time step and dimension of the mesh. Addtionally, several other useful methods are provided. In the provided method `setMaxDistanceAndDirection` the particles velocity is updated to the newly calculated velocity, the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule:
```
Could you expand and provide more details about `Addtionally, several other useful methods are provided`?

gsgall replied:
I have removed this and now it only discusses the other method that is still

simopier commented on doc/content/source/userobjects/ParticleStepperBase.md:
```diff
@@ -0,0 +1,9 @@
+
+\begin{equation}
+  d_\text{max} = \vec{v} \cdot \vec{v} \: dt
+\end{equation}
```
```suggestion
\begin{equation}
  d_\text{max} = \vec{v} \cdot \vec{v} \: dt,
\end{equation}
with ....
```
Please explain the notations.

simopier commented on doc/content/source/userobjects/ParticleStepperBase.md:
```diff
@@ -0,0 +1,9 @@
+  d_\text{max} = \vec{v} \cdot \vec{v} \: dt
+\end{equation}
+
+However, only the components of the velocity data corrisponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is peformed with `ParticleStepperBase::setupStep`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and ...
```
```suggestion
However, only the components of the velocity data corresponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is performed with `ParticleStepperBase::setupStep`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and a time step to be used for the maximum distance calculation to allow flexibility for particle pushing methods.

!syntax parameters /UserObjects/ParticleStepperBase

!syntax inputs /UserObjects/ParticleStepperBase

!syntax children /UserObjects/ParticleStepperBase
```

gsgall replied:
The !syntax commands on this page have been excluded. The particle stepper base is an abstract base class and will never be used directly by end users. These also break the docs when they are added.

simopier commented on include/userobjects/BorisStepper.h:
```diff
@@ -0,0 +1,55 @@
+
+#pragma once
+
+// MOOSE Includes
```
I think this can be removed.
```suggestion
```

simopier commented on include/userobjects/LeapFrogStepper.h:
```diff
@@ -0,0 +1,37 @@
+
+#pragma once
+
+// MOOSE Includes
```
```suggestion
```

simopier commented on include/userobjects/LeapFrogStepper.h:
```diff
@@ -0,0 +1,37 @@
+                         const Real distance = 0) const override final;
+
+private:
+  /// the variable names for each componet of the force vector
```
```suggestion
  /// the variable names for each component of the force vector
```

simopier commented on include/userobjects/PICStudyBase.h:
```diff
@@ -36,19 +42,39 @@ class PICStudyBase : public RayTracingStudy
+  const Elem * _temp_elem;
+  Point _temp_point;
+  Real _temp_distance;
+  Point _temp_velocity;
```
`temp` can sometimes be used for temperature, so using the full name could avoid confusion. 
```suggestion
  const Elem * _temporary_elem;
  Point _temporary_point;
  Real _temporary_distance;
  Point _temporary_velocity;
```

loganharbour replied:
Left a comment in the source; these should be removed

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+
+// MOOSE Includes
```
```suggestion
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+
+  /**
+   * Method available for override to define the rules by which you update the paricles
+   * velocity. The particles velocity data should be updated in this function
```
```suggestion
   * Method available for override to define the rules by which particles' velocity
   * is updated. The particles' velocity data should be updated in this function.
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+
+protected:
+  /**
+   * Method for a simple dimension dependent update of the rays max distance and
```
```suggestion
   * Method for a simple dimension-dependent update of the rays max distance and
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * for a 3D problem all components of the velocity will be used
+   * @param ray the ray whose velocity we want to set
+   * @param v the new velocity that you want the ray to have
+   * @param dt the time step which you are using to set the maximum distance
```
```suggestion
   * @param dt the time step used to set the maximum distance
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * for a 1D problem only the x value of the velocity stored ray data will be used
+   * for a 2D problem only the x and y values of the velocity store in ray data will be used
+   * for a 3D problem all components of the velocity will be used
+   * @param ray the ray whose velocity we want to set
```
```suggestion
   * @param ray the ray whose velocity is being set
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * for a 2D problem only the x and y values of the velocity store in ray data will be used
+   * for a 3D problem all components of the velocity will be used
+   * @param ray the ray whose velocity we want to set
+   * @param v the new velocity that you want the ray to have
```
```suggestion
   * @param v the new ray velocity
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * Used for sampling each component of a finite element field variable
+   * The field will be sampled at the point of the ray passed to the method
+   * @param field_samplers the sampler utilities set up to sample each component of the force field
+   * @param ray the current ray you are updating
```
Please avoid using `you`, and `we` in the comments. 
```suggestion
   * @param ray the current ray
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * @param v the current velocity of the particle
+   * @param F the value of the field at the point the particle exists
+   * @param q_m_ratio the charge to mass ratio of the current particle
+   * @param dt the time step through which you are accelerating the particle
```
```suggestion
   * @param dt the time step
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   * @param dt the time step through which you are accelerating the particle
+   */
+  Point linearImpulse(const Point & v, const Point & F, const Real q_m_ratio, const Real dt) const;
+  /// the dimension of the mesh being using used for dimension dependent velocity updates
```
```suggestion
  /// the dimension of the mesh being using used for dimension-dependent velocity updates
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+   */
+  Point linearImpulse(const Point & v, const Point & F, const Real q_m_ratio, const Real dt) const;
+  /// the dimension of the mesh being using used for dimension dependent velocity updates
+  const Real _dim;
```
More explicit names help avoid confusion and make code easier to read. 
Disregard if `_dim` is commonly used in ray tracing, but keep this general guidance in mind. 
```suggestion
  const Real _mesh_dimension;
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+      "Electromagnetic particle stepper method implementing the Boris Algorithm");
+  params.addRequiredParam<std::vector<VariableName>>(
+      "efield_components",
+      "A list of 3 variables which represent the 3 componets of the electric field");
```
```suggestion
      "A list of 3 variables which represent the 3 components of the electric field");
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+      "A list of 3 variables which represent the 3 componets of the electric field");
+  params.addRequiredParam<std::vector<VariableName>>(
+      "bfield_components",
+      "A list of 3 variables which represent the 3 componets of the electric field");
```
```suggestion
      "A list of 3 variables which represent the 3 components of the electric field");
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+  if (_efield_vars.size() != 3)
+    mooseError("BorisStepper with name ",
+               name(),
+               ": You must provide 3 componets representing the electric field!");
```
```suggestion
               ": You must provide 3 components representing the electric field!");
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+  if (_bfield_vars.size() != 3)
+    mooseError("BorisStepper with name ",
+               name(),
+               ": You must provide 3 componets representing the magnetic field!");
```
```suggestion
               ": You must provide 3 components representing the magnetic field!");
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+
+  Real dt = _dt;
+  // if it is the particles first step we need to take a half step
+  // other wise we take a full step
```
```suggestion
  // otherwise we take a full step
```

simopier commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+  // other wise we take a full step
+  if (distance == 0)
+    dt /= 2;
+  // lets sample the fields at the ray location so we can upate the velocity properly
```
```suggestion
  // let's sample the fields at the ray location so we can update the velocity properly
```

simopier commented on src/userobjects/LeapFrogStepper.C:
```diff
@@ -0,0 +1,58 @@
+                             "the velocity and position are updated with a 1/2 dt offset.");
+  params.addRequiredParam<std::vector<VariableName>>(
+      "field_components",
+      "A list of 3 variables which represent the 3 componets of the force field acting on "
```
```suggestion
      "A list of 3 variables which represent the 3 components of the force field acting on "
```

simopier commented on src/userobjects/LeapFrogStepper.C:
```diff
@@ -0,0 +1,58 @@
+  if (_field_vars.size() != 3)
+    mooseError("LeapFrogStepper with name ",
+               name(),
+               ": You must provide 3 componets representing the force field!");
```
```suggestion
               ": You must provide 3 components representing the force field!");
```

simopier commented on src/userobjects/PICStudyBase.C:
```diff
@@ -69,16 +72,23 @@ PICStudyBase::reinitializeParticles()
-    const auto start_point = ray->currentPoint();
+    _temp_elem = ray->currentElem();
+    _temp_point = ray->currentPoint();
+    _temp_distance = ray->distance();
```
```suggestion
    _temporary_elem = ray->currentElem();
    _temporary_point = ray->currentPoint();
    _temporary_distance = ray->distance();
```

loganharbour replied:
Don't bother with using shared state here.

```
const auto elem = ray->currentElem();
const auto point = ray->currentPoint();
const auto distance = ray->distance();
```

The overhead for constructing these is minimal. In most cases, the compiler is smarter than you and will move them outside of the loop and reuse them. This is a micro-optimization that just introduces confusion for little gain

gsgall replied:
Thank you for the explanation! I have made this change.

simopier commented on src/userobjects/PICStudyBase.C:
```diff
@@ -69,16 +72,23 @@ PICStudyBase::reinitializeParticles()
+    _temp_point = ray->currentPoint();
+    _temp_distance = ray->distance();
 
+    getVelocity(*ray, _temp_velocity);
```
```suggestion
    getVelocity(*ray, _temporary_velocity);
```

simopier commented on src/userobjects/PICStudyBase.C:
```diff
@@ -69,16 +72,23 @@ PICStudyBase::reinitializeParticles()
+                       ray->data()[_charge_index] / ray->data()[_mass_index],
+                       _temp_distance);
+
+    setVelocity(*ray, _temp_velocity);
```
```suggestion
    ray->setStart(_temporary_point, _temporary_elem);
    _stepper.setupStep(*ray,
                       _temporary_velocity,
                       ray->data()[_charge_index] / ray->data()[_mass_index],
                       _temporary_distance);

    setVelocity(*ray, _temporary_velocity);
```

simopier commented on src/userobjects/ParticleStepperBase.C:
```diff
@@ -0,0 +1,74 @@
+  auto params = GeneralUserObject::validParams();
+  params.addClassDescription("Base class for ParticleStepper. Provides the basic implementation"
+                             "for dimensional dependent velocity updating."
+                             "And the ability to sample vector fields for use in a paritcle step");
```
```suggestion
                             "And the ability to sample vector fields for use in a particle step");
```

simopier commented on src/userobjects/ParticleStepperBase.C:
```diff
@@ -0,0 +1,74 @@
+  // max distance is v^2 dt
+  const auto max_dist = std::sqrt(velocity * velocity) * dt;
+
+  ray.setStartingMaxDistance(max_dist);
```
```suggestion
  const auto max_distance = std::sqrt(velocity * velocity) * dt;

  ray.setStartingMaxDistance(max_distance);
```

simopier reviewed (changes requested):
Thank you for this contribution!

A partial first round of reviews (I did not make it to the tests yet). 
It looks like a lot of comments, but these are generally minor since you submitted a very good base.  

EDIT: by the way @gsgall , Casey taught me something I'm now sharing every chance I get: if you go to the `file changes`, you can submit commits in batches from suggestions. 

Let me know if you have any questions.

simopier commented:
I just looked at the test coverage, @gsgall: 
- you'll want to add tests activating the error from `if (******.size() != 3)` check. 
- Also, it looks like the `unused methods` in `ParticleStepperBase` are indeed unused. Do they need to stay there? (@loganharbour)
-  it also look like the following lines are not hit:
``` 
const std::vector<std::shared_ptr<Ray>> &
PICStudyBase::getBankedRays() const
{
  return _banked_rays;
}
```
What are they for?

loganharbour commented on include/userobjects/PICStudyBase.h:
```diff
@@ -23,8 +23,14 @@ class PICStudyBase : public RayTracingStudy
+   * Method for getting the rays after they have finished tracing
+   * useful for looking at the rays data if needed by another object
+   */
+  const std::vector<std::shared_ptr<Ray>> & getBankedRays() const;
```
This isn't used?

gsgall replied:
That is left over from when was using a custom post processor object for the test cases.I just forgot to remove it when I moved to exodiffs for the tests. I'll go ahead and remove it.

loganharbour commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+  ///@{
+  virtual void initialize() override {}
+  virtual void finalize() override {}
+  virtual void execute() override {}
```
```suggestion
  virtual void initialize() override final {}
  virtual void finalize() override final {}
  virtual void execute() override final {}
```

With this, you'll get an error if you try to override in a derived class. Just adds more intent

loganharbour commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+               ": You must provide 3 componets representing the electric field!");
+
+  if (_bfield_vars.size() != 3)
+    mooseError("BorisStepper with name ",
```
paramError

loganharbour commented on src/userobjects/BorisStepper.C:
```diff
@@ -0,0 +1,84 @@
+    _bfield_vars(getParam<std::vector<VariableName>>("bfield_components"))
+{
+  if (_efield_vars.size() != 3)
+    mooseError("BorisStepper with name ",
```
First, you don't need to add context to mooseErrors within a MooseObject. They already have context. The errors look something like this:

```
*** ERROR **

/path/to/object/in/input.i:5.4
The following error occurred in the object "object_name" of type BorisStepper:

<your error here>
```

Secondly, these should be paramErrors, because they're attributed directly to a parameter:

```
paramError("efield_components, "Must provide three components");
```

gsgall replied:
This change was made in both the `BorisStepper` and the `LeapfrogStepper` objects

loganharbour commented on src/userobjects/LeapFrogStepper.C:
```diff
@@ -0,0 +1,58 @@
+    _field_vars(getParam<std::vector<VariableName>>("field_components"))
+{
+  if (_field_vars.size() != 3)
+    mooseError("LeapFrogStepper with name ",
```
paramError

loganharbour commented on src/userobjects/LeapFrogStepper.C:
```diff
@@ -0,0 +1,58 @@
+    return;
+  }
+  v = linearImpulse(v, F, q_m_ratio, _dt);
+  setMaxDistanceAndDirection(ray, v, _dt);
```
```suggestion
  const auto dt = distance == 0 ? _dt / 2 : _dt;
  v = linearImpulse(v, F, q_m_ratio, dt);
  setMaxDistanceAndDirection(ray, v, dt);
```

loganharbour commented on src/userobjects/PICStudyBase.C:
```diff
@@ -90,8 +100,24 @@ PICStudyBase::postExecuteStudy()
+{
+  v(0) = ray.data()[_v_x_index];
+  v(1) = ray.data()[_v_y_index];
+  v(2) = ray.data()[_v_z_index];
```
Not a requirement, but in the future do this:

```
  v(0) = ray.data(_v_x_index);
  v(1) = ray.data(_v_y_index);
  v(2) = ray.data(_v_z_index);
```

The data(std::size_t index) method has bounds checking in debug

gsgall replied:
This change was made in both the `getVelocity` and `setVelocity` methods

loganharbour commented on src/userobjects/ParticleStepperBase.C:
```diff
@@ -0,0 +1,74 @@
+ParticleStepperBase::setMaxDistanceAndDirection(Ray & ray, const Point & v, const Real dt) const
+{
+
+  if (v.absolute_fuzzy_equals(Point(0, 0, 0)))
```
Comment for why?

loganharbour commented on test/tests/userobjects/boris_stepper/cyclotron_motion.i:
```diff
@@ -0,0 +1,164 @@
+[]
+
+[Kernels]
+  [Ex_null]
```
If you're trying to avoid the errors about not having kernels here, you can do

`Problem/kernel_coverage_check=false` 

and can remove these kernels.

gsgall replied:
Thank you telling me about this!

This change was made in all of the new test cases.

loganharbour reviewed (changes requested):
All pretty minor here.

loganharbour commented:
>  it also look like the following lines are not hit:

Heh @simopier noticed without even looking at the coverage 😬

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>
Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

gsgall committed:
Addressing most of the required code changes

- the getBankedRays() method was removed from the PICStudyBase
- some temporary variables in the reinitialize method were removed for clarity
- the temporary_velocity variable was renamed to _temporary_velocity

gsgall commented:
> > it also look like the following lines are not hit:
> 
> Heh @simopier noticed without even looking at the coverage 😬

Sorry about that. I was using that method when I had written a test that output rayData into csv with a custom post processor object. I forgot to remove it when I transitioned the tests to be exodiffs because the post processor test was not parallel consistent.

gsgall committed:
Updating documentation

gsgall commented:
@loganharbour @simopier This should be good to go.

loganharbour commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,77 @@
+#include "Ray.h"
+#include "GeneralUserObject.h"
+#include "VariableSampler.h"
+#include "RayTracingStudy.h"
```
```suggestion
#include "GeneralUserObject.h"

#include "VariableSampler.h"

class Ray;
```

loganharbour reviewed (commented):
My last comment. Content with everything else

gsgall committed:
Update include/userobjects/ParticleStepperBase.h

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

simopier commented on doc/content/source/userobjects/BorisStepper.md:
```diff
@@ -0,0 +1,127 @@
+  \frac{\Delta t}{2}.
+\end{equation}
+
+The implementation of the Boris algorithm was verified using several single particle motion tests: constant electric field, cyclotron motion ([/cyclotron_motion.i]), and $\vec{E} \times \vec{B}$ drift motion ([/e_cross_b.i]).
```
Are you adding the constant electric field case to this PR? I think you should since it's in the paper and people might be looking for it, and I think you should link the input in here like we did for the cyclotron motion test, for example.

gsgall replied:
Those tests were in the leapfrog stepper section of tests since Boris with no Magnetic field is the same as the leap frog stepper. But I have added these tests to the Boris stepper suite as well for good measure and updated the documentation to link to them as well.

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+
+!syntax description /UserObjects/LeapFrogStepper
+
+When charged particles are only subjected to an electric field a standard particle stepping scheme called the leapfrog scheme is used to accurately calculate the motion of these particles through the field. This scheme can repoduce particle paths exactly, to machine precision in the case of a constant uniform electric field. More information about this particle stepping scheme can be found in [!c ...
```
```suggestion
When charged particles are only subjected to an electric field, a standard particle stepping scheme called the leapfrog scheme can be used to accurately calculate the motion of these particles through the field. This scheme can reproduce particle paths exactly, to machine precision, in the case of a constant uniform electric field. More information about this particle stepping scheme can be found in [!cite](Birdsall_Langdon_1991).
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+  \vec{v}_{1/2} = \vec{v}_0 + \frac{q}{m} \vec{F}_0 \frac{dt}{2}
+\end{equation}
+
+Where $\vec{v}_{0}$ is the initial particle velocity, $q$, and $m$ are the particles charge and mass, $\vec{F}_0$ is the initial force field the particle is subject to, and $\vec{v}_{1/2}$ is the particle velocity 1/2 of a time step after the simulation start time.
```
```suggestion
where $\vec{v}_{0}$ is the initial particle velocity, $q$, and $m$ are the particles charge and mass, $\vec{F}_0$ is the initial force field the particle is subject to, and $\vec{v}_{1/2}$ is the particle velocity 1/2 of a time step after the simulation start time.
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+  \vec{r}_{1} = \vec{r}_0 + \vec{v}_{1/2} dt
+\end{equation}
+
+Where $\vec{r}_{1}$ is the particles' position after the first time step has been taken and $\vec{r}_0$ is the initial position of the particle. On the $n^\text{th}$ step after the initial step both the particles' velocity and postition are updated with a full time step
```
```suggestion
where $\vec{r}_{1}$ is the particles' position after the first time step has been taken and $\vec{r}_0$ is the initial position of the particle. On the $n^\text{th}$ step after the initial step both the particles' velocity and postition are updated with a full time step
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+  \vec{v}_{n + 1/2} = \vec{v}_{n - 1/2} + \frac{q}{m} \vec{F} dt
+\end{equation}
+
+Where $\vec{v}_{n + 1/2}$ is the particles' velocity at a time which is 1/2 of a time step after the time at time step $n$, and $\vec{v}_{n - 1/2}$ is the particles' velocity at a time which is 1/2 of a time step before the time at time step $n$. Then the new particle position is updated using the velocity from 1/2 step in the future.
```
```suggestion
where $\vec{v}_{n + 1/2}$ is the particles' velocity at a time which is 1/2 of a time step after the time at time step $n$, and $\vec{v}_{n - 1/2}$ is the particles' velocity at a time which is 1/2 of a time step before the time at time step $n$. Then the new particle position is updated using the velocity from 1/2 step in the future.
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+  \vec{r}_{n} = \vec{r}_{n-1} + \vec{v}_{n - 1/2} dt
+\end{equation}
+
+Where $\vec{r}_{n}$ is the particles position at the time corrisponding to the $n^\text{th}$ time step and $\vec{r}_{n-1}$ is the particles position on the previous time step.
```
```suggestion
where $\vec{r}_{n}$ is the particles position at the time corresponding to the $n^\text{th}$ time step and $\vec{r}_{n-1}$ is the particles position on the previous time step.
```

simopier commented on doc/content/source/userobjects/LeapFrogStepper.md:
```diff
@@ -0,0 +1,47 @@
+
+Where $\vec{r}_{n}$ is the particles position at the time corrisponding to the $n^\text{th}$ time step and $\vec{r}_{n-1}$ is the particles position on the previous time step.
+
+The implementation of the Leap frog scheme was verified using single particle motion tests: parallel acceleration in an electric field ([/parallel_acceleration.i]), and projectile motion ([/projectile_motion.i]).
```
```suggestion
The implementation of the leap frog scheme was verified using single particle motion tests: parallel acceleration in an electric field ([/parallel_acceleration.i]), and projectile motion ([/projectile_motion.i]).
```

simopier replied:
Thanks for the references to the inputs!

simopier commented on doc/content/source/userobjects/ParticleStepperBase.md:
```diff
@@ -0,0 +1,14 @@
+# ParticleStepperBase
+
+`ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particles direction and maximum distance based on the velocity stored in ray data, the current time step and dimension of the mesh. In the provided method `setMaxDistanceAndDirection` the direction is set based on the dimension of the mesh, and the max ...
```
```suggestion
`ParticleStepperBase` is the abstract class form which all steppers should be derived. `ParticleStepperBase` provides the basic method to update a particle's direction and maximum distance based on the velocity stored in ray data, the current time step, and dimension of the mesh. In the provided method `setMaxDistanceAndDirection` the direction is set based on the dimension of the mesh, and the maximum distance is set with the following rule:
```

simopier commented on doc/content/source/userobjects/ParticleStepperBase.md:
```diff
@@ -0,0 +1,14 @@
+  d_\text{max} = \vec{v} \cdot \vec{v} \: dt,
+\end{equation}
+
+Where $d_\text{max}$ is maximum distance the particle is allowed to travel during the nex time step, and $\vec{v}$ is the particles' velocity.
```
```suggestion
where $d_\text{max}$ is maximum distance the particle is allowed to travel during the next time step, and $\vec{v}$ is the particles' velocity.
```

simopier commented on include/userobjects/ParticleStepperBase.h:
```diff
@@ -0,0 +1,78 @@
+  ParticleStepperBase(const InputParameters & parameters);
+
+  static InputParameters validParams();
+  /**
+   * Unused methods
+   */
+  ///@{
+  virtual void initialize() override final {}
+  virtual void finalize() override final {}
+  virtual void execute() override final {}
+  ///@}
```
I don't get why we needed these methods since they are unused.

gsgall replied:
These are virtual methods in the abstract base class UserObject from which GeneralUserObject is derived and when creating a non-abstract child class from an abstract base class all of the virtual methods must be given an implementation. So we have to override them even if we are not using them because the ParticleStepperBase is not an abstract class.

simopier commented on test/include/userobjects/TestSimpleStepper.h:
```diff
@@ -0,0 +1,29 @@
+
+#pragma once
+
+// MOOSE Includes
```
```suggestion
```

simopier commented on test/include/userobjects/TestSimpleStepper.h:
```diff
@@ -0,0 +1,29 @@
+  static InputParameters validParams();
+
+  /**
+   * A simple particle stepper which will not modify the velocity of the paritcle
```
```suggestion
   * A simple particle stepper which will not modify the velocity of the particle
```

simopier commented on test/src/userobjects/TestPICStudyUserDefinedRays.C:
```diff
@@ -25,14 +25,17 @@ TestPICStudyUserDefinedRays::validParams()
       "The direction(s) that the ray(s) start in (does not need to be normalized)");
-
+  params.addParam<Real>("mass", 0, "The mass of the particles used for during a test");
+  params.addParam<Real>("charge", 0, "The charge of the particles used for during a test");
```
```suggestion
  params.addParam<Real>("mass", 0, "The mass of the particles used for a test");
  params.addParam<Real>("charge", 0, "The charge of the particles used for a test");
```

simopier commented on test/tests/userobjects/leapfrog_stepper/tests:
```diff
@@ -0,0 +1,48 @@
+               projectile_motion_rays.e-s0008
+               projectile_motion_rays.e-s0009'
+    allow_test_objects = true
+    requirement = 'The system shall be capable of applying a linear impluse from a force field perpendicular to a particles initial velocity'
```
```suggestion
    requirement = 'The system shall be capable of applying a linear impulse from a force field perpendicular to a particle's initial velocity'
```

simopier commented on test/tests/userobjects/leapfrog_stepper/tests:
```diff
@@ -0,0 +1,48 @@
+               parallel_acceleration_rays.e-s0008
+               parallel_acceleration_rays.e-s0009'
+    allow_test_objects = true
+    requirement = 'The system shall be capable of applying a linear impluse from a force field parallel to a particles velocity'
```
```suggestion
    requirement = 'The system shall be capable of applying a linear impulse from a force field parallel to a particle's velocity'
```

simopier reviewed (changes requested):
Thank you for these improvements, @gsgall. I have a few more minor comments and two suggestions:
- could you add a test case for the constant electric field case?
- I encourage you to add a reference to your submitted PBNC paper in the relevant documentation pages. This will improve its visibility. You can lark it as `submitted`, and we'll update this when it gets published.

gsgall committed:
Adding efield tests for boris stepper

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

gsgall committed:
Update doc/content/source/userobjects/ParticleStepperBase.md

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

gsgall committed:
resolving merge issue

gsgall committed:
Update test/tests/userobjects/leapfrog_stepper/tests

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

gsgall committed:
Adding PBNC conference reference

gsgall committed:
Adding the PNBC reference

gsgall committed:
Merge branch 'devel' of https://github.com/gsgall/fenix into devel

gsgall commented:
> Thank you for these improvements, @gsgall. I have a few more minor comments and two suggestions:
> 
> * could you add a test case for the constant electric field case?
> * I encourage you to add a reference to your submitted PBNC paper in the relevant documentation pages. This will improve its visibility. You can lark it as `submitted`, and we'll update this when it gets published.

I have added these test cases for the BorisStepper I also left a comment about them on your review requests where you ask for them as well. 

Additionally, the paper reference has been added to the BorisStepper documentation page.

gsgall committed:
fixing the apostrophe's that broke the tests

simopier reviewed (approved):
Thanks @gsgall, this looks great to me. 

@loganharbour, could we add the json file capabilities you mentioned to make the gold file more human readable? Maybe in a follow up PR if it isn't ready to go in yet.

loganharbour commented:
> @loganharbour, could we add the json file capabilities you mentioned to make the gold file more human readable? Maybe in a follow up PR if it isn't ready to go in yet.

Yeah, this'll be a future effort. It's on a list :)
